include/inexor/vulkan-renderer/wrapper/query_pool.hpp

```diff
@@ -0,0 +1,100 @@
+        VK_QUERY_PIPELINE_STATISTIC_INPUT_ASSEMBLY_PRIMITIVES_BIT,
+        VK_QUERY_PIPELINE_STATISTIC_VERTEX_SHADER_INVOCATIONS_BIT,
+        VK_QUERY_PIPELINE_STATISTIC_CLIPPING_INVOCATIONS_BIT,
+        VK_QUERY_PIPELINE_STATISTIC_CLIPPING_PRIMITIVES_BIT,
+        VK_QUERY_PIPELINE_STATISTIC_FRAGMENT_SHADER_INVOCATIONS_BIT,
+        VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_INVOCATIONS_BIT,                // requires geometry shaders
+        VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_PRIMITIVES_BIT,                 // requires geometry shaders
+        VK_QUERY_PIPELINE_STATISTIC_TESSELLATION_CONTROL_SHADER_PATCHES_BIT,        // requires tesselation shaders
+        VK_QUERY_PIPELINE_STATISTIC_TESSELLATION_EVALUATION_SHADER_INVOCATIONS_BIT, // requires tesselation shaders
+        VK_QUERY_PIPELINE_STATISTIC_COMPUTE_SHADER_INVOCATIONS_BIT};
+
+    [[nodiscard]] static std::string get_pipeline_stats_flag_bit_name(VkQueryPipelineStatisticFlagBits bit);
+
+    /// @brief Validates every specified VkQueryPipelineStatisticFlagBits into one VkQueryPipelineStatisticFlags.
+    /// Some VkQueryPipelineStatisticFlagBits values require special checks (tesselation shaders for example).
+    /// @return A vector which contains the valid Vulkan pipeline query statistics flag bits.
+    [[nodiscard]] std::vector<VkQueryPipelineStatisticFlagBits>
+    validate_pipeline_stats_flag_bits(const std::vector<VkQueryPipelineStatisticFlagBits> &pipeline_stats_flag_bits);
+
+public:
+    /// @brief Construct a Vulkan Query Pool using the default pipeline statistics.
+    /// @param device The const reference to a device RAII wrapper instance.
+    /// @param name The internal name.
+    QueryPool(const Device &device, const std::string &name);
+
+    /// Call <a href="https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#vkCreateQueryPool">
+    /// vkCreateQueryPool</a>
+    /// @param device The device wrapper.
+    /// @param name The internal name of this performance query.
+    /// @param pipeline_stats_flag_bits The enabled <a
+    /// href="https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VkQueryPipelineStatisticFlagBits">
+    /// Vulkan pipeline statistics flags</a>.
+    QueryPool(const Device &device, const std::string &name,
+              const std::vector<VkQueryPipelineStatisticFlagBits> &pipeline_stats_flag_bits);
+
+    QueryPool(const QueryPool &) = delete;
+    QueryPool(QueryPool &&) = delete;
+    ~QueryPool();
+
+    QueryPool operator=(const QueryPool &) = delete;
+    QueryPool operator=(QueryPool &&) = delete;
+
+    [[nodiscard]] const wrapper::Device &device() const noexcept {
+        return m_device;
+    }
+
+    /// Call <a
+    /// href="https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#vkCmdResetQueryPool">
+    /// vkCmdResetQueryPool</a>
+    void reset(VkCommandBuffer cmd_buffer) const;
```

Comment:
Maybe would be a bit nicer if these took in a `const CommandBuffer &` instead? Then no need to `assert` since it can't be in an invalid state :)

Reply:
Good idea. Done.